modules/git/repo_compare.go

```diff
@@ -233,72 +233,63 @@ func parseDiffStat(stdout string) (numFiles, totalAdditions, totalDeletions int,
 	return numFiles, totalAdditions, totalDeletions, err
 }
 
-// GetDiffOrPatch generates either diff or formatted patch data between given revisions
-func (repo *Repository) GetDiffOrPatch(base, head string, w io.Writer, patch, binary bool) error {
-	if patch {
-		return repo.GetPatch(base, head, w)
+func parseCompareArgs(compareArgs string) (args []string) {
+	parts := strings.Split(compareArgs, "...")
+	if len(parts) == 2 {
+		return []string{compareArgs}
 	}
-	if binary {
-		return repo.GetDiffBinary(base, head, w)
+	parts = strings.Split(compareArgs, "..")
+	if len(parts) == 2 {
+		return parts
 	}
-	return repo.GetDiff(base, head, w)
+	parts = strings.Fields(compareArgs)
+	if len(parts) == 2 {
+		return parts
+	}
+
+	return nil
 }
 
 // GetDiff generates and returns patch data between given revisions, optimized for human readability
-func (repo *Repository) GetDiff(base, head string, w io.Writer) error {
+func (repo *Repository) GetDiff(compareArgs string, w io.Writer) error {
+	args := parseCompareArgs(compareArgs)
+	if len(args) == 0 {
+		return fmt.Errorf("invalid compareArgs: %s", compareArgs)
+	}
 	stderr := new(bytes.Buffer)
-	err := NewCommand(repo.Ctx, "diff", "-p").AddDynamicArguments(base + "..." + head).
+	return NewCommand(repo.Ctx, "diff", "-p").AddDynamicArguments(args...).
 		Run(&RunOpts{
 			Dir:    repo.Path,
 			Stdout: w,
 			Stderr: stderr,
 		})
-	if err != nil && bytes.Contains(stderr.Bytes(), []byte("no merge base")) {
-		return NewCommand(repo.Ctx, "diff", "-p").AddDynamicArguments(base, head).
-			Run(&RunOpts{
-				Dir:    repo.Path,
-				Stdout: w,
-			})
-	}
```

Comment:
Are you sure `git diff a...b` is the same as `git diff a b`?

Reply:
`git diff a..b` is equal to `git diff a b` which is different from `git diff a...b`. In this function, the old logic is to try `git diff a...b` first and if it's failed, then try `git diff a..b`. This is not right for those users requests a `git diff a...b` compare. So that, the new logic removed the guess logic. Just follow the user's request from UI or API, it will fail if the compare fails.

Reply:
But are you sure the changed logic is right for most end users?

Your `parseCompareArgs` loses the `...` or `..` information.

Reply:
All these functions are only used in `DownloadDiffOrPatch(ctx context.Context, pr *issues_model.PullRequest, w io.Writer, patch, binary bool)`. Which is a comparison for the pull request's base branch and head branch. For a pull request, we should always use `...` rather than `..`. This is also what GitHub did for the pull request. ref: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#about-three-dot-comparison-on-github

Reply:
Then should `DownloadDiffOrPatch` use `...` or `..`?

Reply:
It should use `...` because it's from a pull request.

Reply:
`` DownloadDiffOrPatch -> GetPatch -> parseCompareArgs -> AddDynamicArguments(args...) ``

Then `git diff a b`?

Reply:
`parseCompareArgs` will keep `a...b` as `a...b` and convert `a..b` to `a` and `b`. ref https://github.com/go-gitea/gitea/pull/32817/files#diff-e95e1cf24a888677dc1289a1e90bec1c0deeba77259df10d2bd81d426e6af1f7R239 and also there are some tests for that.

Reply:
OK, I see the key point ` []string{compareArgs}` is really counter-intuitive